mlflow/pyfunc/__init__.py

```diff
@@ -939,7 +1132,53 @@ def predict(*args):
         else:
             return result[result.columns[0]]
 
-    udf = pandas_udf(predict, result_type)
+    def predict(iterator):
+        import requests
+        from mlflow.pyfunc.scoring_server.client import ScoringServerClient
+
+        # Note: this is a pandas udf function in iteration style, which takes an iterator of
+        # tuple of pandas.Series and outputs an iterator of pandas.Series.
+        if not use_nfs and not spark_in_local_mode:
+            model_on_executor, local_model_path_on_executor = SparkModelCache.get_or_load(model_archive_path)
+            virtual_env_python_bin_on_executor = _setup_restored_python_env(model_py_version, req_file_path)
+        else:
+            model_on_executor, local_model_path_on_executor = SparkModelCache.get_or_load_from_nfs_dir(local_model_path)
+            virtual_env_python_bin_on_executor = virtual_env_python_bin
+
+        server_port = find_free_port()
+        # Command: __pyfunc_model_local_path__=... gunicorn -b 127.0.0.1:{port} -w 1 --timeout=500 -- mlflow.pyfunc.scoring_server.wsgi:app
+        server_proc_env = os.environ.copy()
+        server_proc_env[scoring_server._SERVER_MODEL_LOCAL_PATH] = local_model_path_on_executor
+        gunicorn_bin_path = os.path.join(os.path.dirname(virtual_env_python_bin_on_executor), "gunicorn")
+        # launch scoring server
+        scoring_server_proc = subprocess.Popen(
+            [
+                gunicorn_bin_path, "-b", f"127.0.0.1:{server_port}", "-w", "1", "--timeout=500", "--",
+                "mlflow.pyfunc.scoring_server.wsgi:app"
+            ],
+            stdout=sys.stdout, stderr=sys.stderr,
+            env=server_proc_env
+        )
```

Comment:
Could we use `PyFuncBackend.serve()` instead? https://github.com/mlflow/mlflow/blob/5e2c868d69aad968dd6eaebf8c8f447e72366140/mlflow/pyfunc/backend.py#L67.

This way, we can support both conda and virtualenv / pyenv. We should use conda by default in MLflow 1.x. Can you work with @harupy to add pyenv / virtualenv support to `PyFuncBackend`? @harupy perhaps it would be helpful to have @WeichenXu123 implement this, since it directly relates to `spark_udf` environment restoration.

Reply:
I considered this, I don't like do setup restored python env inside "PyFuncBackend.serve()", I hope in databricks we can apply the optimization: in driver side directly in the NFS directory setup restored python env (by virtualenv/conda) , and then all executors can use the restored python env. The optimization improve performance very much.

For reusing code, we can do:
* provide set-up virtualenv/conda env helper function
* in serve we can call the helper function above to setup env (if specified argument)
* in spark_udf and in databricks, we call the helper function above to setup env in Driver side in NFS directory. then in `PyFuncBackend.serve()` we skip setting up restored env.

Reply:
Can't we set up the environment using NFS and then call `PyFuncBackend.serve()` after? `PyFuncBackend.serve()` should then see that the environment is already there. We use hashes of the conda environment file to accomplish this for `mlflow models serve`: if a conda environment already exists with a matching hash, we just activate it. We can implement the same thing for virtualenv / pyenv.

Reply:
> We use hashes of the conda environment file to accomplish this for mlflow models serve: if a conda environment already exists with a matching hash, we just activate it

good idea. let me try it.